fvm/src/syscalls/mod.rs

```diff
@@ -41,6 +43,12 @@ pub struct InvocationData<K> {
     /// `last_milligas_available`.
     pub last_milligas_available: i64,
 
+    /// The total size of the memory used by the execution at the beginning of this call.
+    pub start_memory_bytes: usize,
+
+    /// The total size of the memory used by the execution the last time we charged gas for it.
+    pub last_memory_bytes: usize,
```

Comment:
We need to charge on use. After thinking through this, I'm pretty sure the _only_ way to do this will be to charge through instrumentation.

Reply:
Basically:

1. Right now, we charge after the fact. A user could rack up more gas charges than they can afford.
2. If we tried to charge inside the limiter, we'd need to a reference to some form of reference-counted gas counter, which I'd rather not have.
3. If we charge inside instrumentation (before every call to "reserve"), we'd be able to use the current wasm gas counter.

Reply:
Have you seen my comments [here](https://github.com/filecoin-project/ref-fvm/pull/1060#discussion_r1017708956) and [here](https://github.com/filecoin-project/ref-fvm/pull/1039#issuecomment-1307025078)? I posit that we are doing things exactly the way you describe in 3). 

The user cannot rack up more charges than they can afford because [this injection](https://github.com/filecoin-project/fvm-wasm-instrument/blob/master/src/gas_metering/mod.rs#L469) will make it [call the gas function before growing](https://github.com/filecoin-project/fvm-wasm-instrument/blob/master/src/gas_metering/mod.rs#L504-L505).

Reply:
Ah, I'm sorry, there are just too many threads to keep track of here.

You're right, we're already charging. I hadn't realized that that code had already landed in the instrumentation logic. I saw you doing memory things in `charge_for_exec` and got confused.